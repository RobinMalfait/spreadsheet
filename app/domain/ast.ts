import type { Location } from '~/domain/expression'

export enum AstKind {
  CELL = 'CELL',
  RANGE = 'RANGE',
  FUNCTION = 'FUNCTION',
  NUMBER_LITERAL = 'NUMBER_LITERAL',
  STRING_LITERAL = 'STRING_LITERAL',
  BINARY_EXPRESSION = 'BINARY_EXPRESSION',
}

interface Span {
  span: {
    start: number
    end: number
  }
}

export interface AstCell extends Span {
  kind: AstKind.CELL
  name: string
  raw: string
  loc: Location
}

export interface AstCellRange extends Span {
  kind: AstKind.RANGE
  raw: string
  start: AstCell
  end: AstCell
}

export interface AstFunction extends Span {
  kind: AstKind.FUNCTION
  name: string
  args: AST[]
}

interface AstNumberLiteral extends Span {
  kind: AstKind.NUMBER_LITERAL
  value: number
}

interface AstStringLiteral extends Span {
  kind: AstKind.STRING_LITERAL
  value: string
}

interface AstBinaryExpression extends Span {
  kind: AstKind.BINARY_EXPRESSION
  operator: BinaryExpressionOperator
  lhs: AST
  rhs: AST
}

export type AST =
  | AstCell
  | AstCellRange
  | AstFunction
  | AstNumberLiteral
  | AstStringLiteral
  | AstBinaryExpression

export enum BinaryExpressionOperator {
  // Math operators
  EXPONENT = 'EXPONENT',
  MULTIPLY = 'MULTIPLY',
  ADD = 'ADD',
  SUBTRACT = 'SUBTRACT',
  DIVIDE = 'DIVIDE',

  // Comparison operators
  EQUALS = 'EQUALS',
  NOT_EQUALS = 'NOT_EQUALS',
  LESS_THAN = 'LESS_THAN',
  LESS_THAN_EQUALS = 'LESS_THAN_EQUALS',
  GREATER_THAN = 'GREATER_THAN',
  GREATER_THAN_EQUALS = 'GREATER_THAN_EQUALS',
}
