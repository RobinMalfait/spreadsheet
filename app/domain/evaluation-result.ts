export enum EvaluationResultKind {
  ERROR = 'ERROR',
  EMPTY = 'EMPTY',
  NUMBER = 'NUMBER',
  STRING = 'STRING',
  BOOLEAN = 'BOOLEAN',
  DATETIME = 'DATETIME',
}

export type EvaluationResult =
  | { kind: EvaluationResultKind.ERROR; value: string }
  | { kind: EvaluationResultKind.EMPTY; value: string }
  | { kind: EvaluationResultKind.NUMBER; value: number }
  | { kind: EvaluationResultKind.STRING; value: string }
  | { kind: EvaluationResultKind.BOOLEAN; value: boolean }
  | { kind: EvaluationResultKind.DATETIME; value: Date; date: boolean; time: boolean }
