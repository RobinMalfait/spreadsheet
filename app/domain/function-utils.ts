import type { EvaluationResult } from '~/domain/evaluation-result'
import { parse } from '~/domain/signature/parser'
import { tokenize } from '~/domain/signature/tokenizer'
import { validate } from '~/domain/signature/validate'

export function expose<
  T extends EvaluationResult,
  R extends EvaluationResult | EvaluationResult[],
>(signature: string, handle: (...args: T[]) => R): (...args: T[]) => R {
  let sig = parse(tokenize(signature))

  return Object.assign(
    (...args: T[]) => {
      let errors = validate(sig, args)
      if (errors) return errors as R

      // Handle
      return handle(...args) as R
    },
    { signature: sig },
  )
}
