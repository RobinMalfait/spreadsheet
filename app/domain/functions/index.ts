export * from './binary'
export * from './date'
export * from './logic'
export * from './lookup'
export * from './math'
export * from './privileged'
export * from './sequence'
export * from './statistics'
export * from './text'
export * from './types'
