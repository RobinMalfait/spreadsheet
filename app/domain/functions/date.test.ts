import { beforeAll, describe, expect, it, vi } from 'vitest'
import { Spreadsheet } from '~/domain/spreadsheet'
import { visualizeSpreadsheet } from '~/test/utils'

beforeAll(() => {
  vi.setSystemTime(new Date(2013, 0, 21, 8, 15, 20))
})

describe('NOW()', () => {
  it('should error when providing an argument', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: NOW() does not take any arguments
      "
    `)
  })

  it('should return the current date and time', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┐
      │   │ A                   │
      ├───┼─────────────────────┤
      │ 1 │ 2013-01-21 08:15:20 │
      └───┴─────────────────────┘
      "
    `)
  })
})

describe('TODAY()', () => {
  it('should error when providing an argument', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TODAY(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: TODAY() does not take any arguments
      "
    `)
  })

  it('should return the current date and time', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TODAY()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬────────────┐
      │   │ A          │
      ├───┼────────────┤
      │ 1 │ 2013-01-21 │
      └───┴────────────┘
      "
    `)
  })
})

describe('TIME()', () => {
  it('should error when providing an argument', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TIME(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: TIME() does not take any arguments
      "
    `)
  })

  it('should return the current time', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=TIME()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬──────────┐
      │   │ A        │
      ├───┼──────────┤
      │ 1 │ 08:15:20 │
      └───┴──────────┘
      "
    `)
  })
})

describe('DAY(date)', () => {
  it('should error when required argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=DAY()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: DAY() requires a date
      "
    `)
  })

  it('should error when required argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=DAY(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: DAY() requires a date
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=DAY(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: DAY() does not take more than one argument
      "
    `)
  })

  it('should return the day of the given date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=DAY(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬────┐
      │   │ A                   │ B  │
      ├───┼─────────────────────┼────┤
      │ 1 │ 2013-01-21 08:15:20 │ 21 │
      └───┴─────────────────────┴────┘
      "
    `)
  })
})

describe('MONTH(date)', () => {
  it('should error when required argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=MONTH()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: MONTH() requires a date
      "
    `)
  })

  it('should error when required argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=MONTH(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: MONTH() requires a date
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=MONTH(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: MONTH() does not take more than one argument
      "
    `)
  })

  it('should return the month of the given date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=MONTH(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬───┐
      │   │ A                   │ B │
      ├───┼─────────────────────┼───┤
      │ 1 │ 2013-01-21 08:15:20 │ 1 │
      └───┴─────────────────────┴───┘
      "
    `)
  })
})

describe('YEAR(date)', () => {
  it('should error when required argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=YEAR()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: YEAR() requires a date
      "
    `)
  })

  it('should error when required argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=YEAR(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: YEAR() requires a date
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=YEAR(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: YEAR() does not take more than one argument
      "
    `)
  })

  it('should return the year of the given date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=YEAR(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬──────┐
      │   │ A                   │ B    │
      ├───┼─────────────────────┼──────┤
      │ 1 │ 2013-01-21 08:15:20 │ 2013 │
      └───┴─────────────────────┴──────┘
      "
    `)
  })
})

describe('HOUR(date)', () => {
  it('should error when required argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=HOUR()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: HOUR() requires a date
      "
    `)
  })

  it('should error when required argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=HOUR(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: HOUR() requires a date
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=HOUR(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: HOUR() does not take more than one argument
      "
    `)
  })

  it('should return the hour of the given date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=HOUR(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬───┐
      │   │ A                   │ B │
      ├───┼─────────────────────┼───┤
      │ 1 │ 2013-01-21 08:15:20 │ 8 │
      └───┴─────────────────────┴───┘
      "
    `)
  })
})

describe('MINUTE(date)', () => {
  it('should error when required argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=MINUTE()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: MINUTE() requires a date
      "
    `)
  })

  it('should error when required argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=MINUTE(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: MINUTE() requires a date
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=MINUTE(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: MINUTE() does not take more than one argument
      "
    `)
  })

  it('should return the minute of the given date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=MINUTE(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬────┐
      │   │ A                   │ B  │
      ├───┼─────────────────────┼────┤
      │ 1 │ 2013-01-21 08:15:20 │ 15 │
      └───┴─────────────────────┴────┘
      "
    `)
  })
})

describe('SECOND(date)', () => {
  it('should error when required argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SECOND()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SECOND() requires a date
      "
    `)
  })

  it('should error when required argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SECOND(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SECOND() requires a date
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SECOND(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SECOND() does not take more than one argument
      "
    `)
  })

  it('should return the second of the given date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=SECOND(A1)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬────┐
      │   │ A                   │ B  │
      ├───┼─────────────────────┼────┤
      │ 1 │ 2013-01-21 08:15:20 │ 20 │
      └───┴─────────────────────┴────┘
      "
    `)
  })
})

describe('ADD_DAYS(date, days)', () => {
  it('should error when required first argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_DAYS()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_DAYS() requires a date
      "
    `)
  })

  it('should error when required first argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_DAYS(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_DAYS() requires a date
      "
    `)
  })

  it('should error when required second argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_DAYS(NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_DAYS() requires a number of days
      "
    `)
  })

  it('should error when required second argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_DAYS(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_DAYS() requires a number of days
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_DAYS(NOW(), 123, NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_DAYS() does not take more than two arguments
      "
    `)
  })

  it('should add the given number of days to the date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=ADD_DAYS(A1, 5)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬─────────────────────┐
      │   │ A                   │ B                   │
      ├───┼─────────────────────┼─────────────────────┤
      │ 1 │ 2013-01-21 08:15:20 │ 2013-01-26 08:15:20 │
      └───┴─────────────────────┴─────────────────────┘
      "
    `)
  })
})

describe('SUB_DAYS(date, days)', () => {
  it('should error when required first argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_DAYS()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_DAYS() requires a date
      "
    `)
  })

  it('should error when required first argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_DAYS(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_DAYS() requires a date
      "
    `)
  })

  it('should error when required second argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_DAYS(NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_DAYS() requires a number of days
      "
    `)
  })

  it('should error when required second argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_DAYS(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_DAYS() requires a number of days
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_DAYS(NOW(), 123, NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_DAYS() does not take more than two arguments
      "
    `)
  })

  it('should subtract the given number of days from the date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=SUB_DAYS(A1, 5)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬─────────────────────┐
      │   │ A                   │ B                   │
      ├───┼─────────────────────┼─────────────────────┤
      │ 1 │ 2013-01-21 08:15:20 │ 2013-01-16 08:15:20 │
      └───┴─────────────────────┴─────────────────────┘
      "
    `)
  })
})

describe('ADD_HOURS(date, days)', () => {
  it('should error when required first argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_HOURS()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_HOURS() requires a date
      "
    `)
  })

  it('should error when required first argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_HOURS(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_HOURS() requires a date
      "
    `)
  })

  it('should error when required second argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_HOURS(NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_HOURS() requires a number of hours
      "
    `)
  })

  it('should error when required second argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_HOURS(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_HOURS() requires a number of hours
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=ADD_HOURS(NOW(), 123, NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: ADD_HOURS() does not take more than two arguments
      "
    `)
  })

  it('should add the given number of hours to the date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=ADD_HOURS(A1, 8)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬─────────────────────┐
      │   │ A                   │ B                   │
      ├───┼─────────────────────┼─────────────────────┤
      │ 1 │ 2013-01-21 08:15:20 │ 2013-01-21 16:15:20 │
      └───┴─────────────────────┴─────────────────────┘
      "
    `)
  })
})

describe('SUB_DAYS(date, days)', () => {
  it('should error when required first argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_HOURS()')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_HOURS() requires a date
      "
    `)
  })

  it('should error when required first argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_HOURS(123)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_HOURS() requires a date
      "
    `)
  })

  it('should error when required second argument is not passed', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_HOURS(NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_HOURS() requires a number of hours
      "
    `)
  })

  it('should error when required second argument is not the right type', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_HOURS(NOW(), NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_HOURS() requires a number of hours
      "
    `)
  })

  it('should error when passing additional arguments', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=SUB_HOURS(NOW(), 123, NOW())')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬───────┐
      │   │ A     │
      ├───┼───────┤
      │ 1 │ Error │
      └───┴───────┘

      Errors:

      · A1: SUB_HOURS() does not take more than two arguments
      "
    `)
  })

  it('should subtract the given number of hours from the date', () => {
    let spreadsheet = new Spreadsheet()
    spreadsheet.set('A1', '=NOW()')
    spreadsheet.set('B1', '=SUB_HOURS(A1, 5)')

    expect(visualizeSpreadsheet(spreadsheet)).toMatchInlineSnapshot(`
      "
      ┌───┬─────────────────────┬─────────────────────┐
      │   │ A                   │ B                   │
      ├───┼─────────────────────┼─────────────────────┤
      │ 1 │ 2013-01-21 08:15:20 │ 2013-01-21 03:15:20 │
      └───┴─────────────────────┴─────────────────────┘
      "
    `)
  })
})
