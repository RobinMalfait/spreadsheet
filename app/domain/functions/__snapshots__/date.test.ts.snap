// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`@example tests > ADD_DAYS(TODAY(), 7) 1`] = `
"
┌───┬────────────┐
│   │ A          │
├───┼────────────┤
│ 1 │ 2013-01-28 │
└───┴────────────┘
"
`;

exports[`@example tests > ADD_HOURS(NOW(), 8) 1`] = `
"
┌───┬─────────────────────┐
│   │ A                   │
├───┼─────────────────────┤
│ 1 │ 2013-01-21 16:15:20 │
└───┴─────────────────────┘
"
`;

exports[`@example tests > DAY(TODAY()) 1`] = `
"
┌───┬────┐
│   │ A  │
├───┼────┤
│ 1 │ 21 │
└───┴────┘
"
`;

exports[`@example tests > HOUR(NOW()) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 8 │
└───┴───┘
"
`;

exports[`@example tests > MINUTE(NOW()) 1`] = `
"
┌───┬────┐
│   │ A  │
├───┼────┤
│ 1 │ 15 │
└───┴────┘
"
`;

exports[`@example tests > MONTH(TODAY()) 1`] = `
"
┌───┬───┐
│   │ A │
├───┼───┤
│ 1 │ 1 │
└───┴───┘
"
`;

exports[`@example tests > NOW() 1`] = `
"
┌───┬─────────────────────┐
│   │ A                   │
├───┼─────────────────────┤
│ 1 │ 2013-01-21 08:15:20 │
└───┴─────────────────────┘
"
`;

exports[`@example tests > SECOND(NOW()) 1`] = `
"
┌───┬────┐
│   │ A  │
├───┼────┤
│ 1 │ 20 │
└───┴────┘
"
`;

exports[`@example tests > SUB_DAYS(TODAY(), 7) 1`] = `
"
┌───┬────────────┐
│   │ A          │
├───┼────────────┤
│ 1 │ 2013-01-14 │
└───┴────────────┘
"
`;

exports[`@example tests > SUB_HOURS(NOW(), 8) 1`] = `
"
┌───┬─────────────────────┐
│   │ A                   │
├───┼─────────────────────┤
│ 1 │ 2013-01-21 00:15:20 │
└───┴─────────────────────┘
"
`;

exports[`@example tests > TIME() 1`] = `
"
┌───┬──────────┐
│   │ A        │
├───┼──────────┤
│ 1 │ 08:15:20 │
└───┴──────────┘
"
`;

exports[`@example tests > TODAY() 1`] = `
"
┌───┬────────────┐
│   │ A          │
├───┼────────────┤
│ 1 │ 2013-01-21 │
└───┴────────────┘
"
`;

exports[`@example tests > YEAR(TODAY()) 1`] = `
"
┌───┬──────┐
│   │ A    │
├───┼──────┤
│ 1 │ 2013 │
└───┴──────┘
"
`;
